qiskit/aqua/algorithms/single_sample/hhl/hhl.py

```diff
@@ -0,0 +1,328 @@
+        'input_schema': {
+            '$schema': 'http://json-schema.org/schema#',
+            'id': 'hhl_schema',
+            'type': 'object',
+            'properties': {
+            },
+            'additionalProperties': False
+        },
+        'problems': ['linear_system'],
+        'depends': [
+            {'pluggable_type': 'initial_state',
+             'default': {
+                     'name': 'CUSTOM',
+                }
+             },
+            {'pluggable_type': 'eigs',
+             'default': {
+                     'name': 'EigsQPE',
+                     'num_ancillae': 6,
+                     'num_time_slices': 50,
+                     'expansion_mode': 'suzuki',
+                     'expansion_order': 2
+                }
+             },
+            {'pluggable_type': 'reciprocal',
+             'default': {
+                     'name': 'Lookup'
+                }
+             }
+        ],
+    }
+
+    def __init__(
+            self,
+            matrix=None,
+            vector=None,
+            eigs=None,
+            init_state=None,
+            reciprocal=None,
+            num_q=0,
+            num_a=0
+    ):
+        """
+        Constructor.
+
+        Args:
+            matrix (array): the input matrix of linear system of equations
+            vector (array): the input vector of linear system of equations
+            eigs (Eigenvalues): the eigenvalue estimation instance
+            init_state (InitialState): the initial quantum state preparation
+            reciprocal (Reciprocal): the eigenvalue reciprocal and controlled rotation instance
+            num_q (int): number of qubits required for the matrix Operator instance
+            num_a (int): number of ancillary qubits for Eigenvalues instance
+        """
+        super().__init__()
+        super().validate(locals())
+        self._matrix = matrix
+        self._vector = vector
+        self._eigs = eigs
+        self._init_state = init_state
+        self._reciprocal = reciprocal
+        self._num_q = num_q
+        self._num_a = num_a
+        self._circuit = None
+        self._io_register = None
+        self._eigenvalue_register = None
+        self._ancilla_register = None
+        self._success_bit = None
+        self._ret = {}
+
+    @classmethod
+    def init_params(cls, params, algo_input):
+        """Initialize via parameters dictionary and algorithm input instance
+
+        Args:
+            params: parameters dictionary
+            algo_input: LinearSystemInput instance
+        """
+        if algo_input is None:
+            raise AquaError("LinearSystemInput instance is required.")
+
+        matrix = algo_input.matrix
+        vector = algo_input.vector
+        if not isinstance(matrix, np.ndarray):
+            matrix = np.array(matrix)
+        if not isinstance(vector, np.ndarray):
+            vector = np.array(vector)
+
+        # extend vector and matrix for nonhermitian/non 2**n size matrices
```

Comment:
could we postpone this change in the constructor?  users might input a 3x3 matrix in the constructor as well, in this case, we also need to extend it to a 4x4 matrix. 
Furthermore, will hhl record the extended dimension? In the final result, will hhl prune those extra dimensions?